contents/blog/open-source-telemetry-ethical.md

```diff
@@ -0,0 +1,102 @@
+
+### Utilitarian-based ethics
+
+Moral behavior means opting for the greatest common good. An act is moral if it increases the happiness or decreases the misery of the greatest number of people. The results are the measure of moral behavior, not the intent.
+
+Stealing those chips? Go for it.
+
+## So, is it ok for open source projects to track users?
+
+Open source projects are zero expectation - they may contain bugs, security risks, or have confusingly outdated documentation. Remember that time you lost hours on a project that just didn't work in the first place?
+
+The fact it's acceptable for open source software not to work is a _great_ thing - it makes it easier for anyone to contribute. Mistakes, or half-baked projects included. You can't have innovation without a lot of people being able to contribute and make mistakes, or give up.
+
+From a utilitarian perspective, we believe that tracking the minimum data possible to build useful technology, in a way that doesn't share that data with 3rd parties, means that you have the right approach to product analytics in your open source project.
+
+From a duty-based perspective, we'd argue that transparency is key to covering intent. Since most people won't assume they're being tracked, that means being upfront in the docs and not relying on people digging through your code.
+
+## Sorting the good from the questionable
+
+First of all, open source is exactly that. That's a big step up from SAAS - you can immediately inspect the code or run it on a computer with no internet connection if needed. However, the precise function of the telemetry has user implications.
+
+### What is being tracked
+
+The first step is to think about how much, if anything, a project really needs to track.
+
+#### Nothing
+
+Nada, zilch. 
+
+If an open source project exists for fun, for one person only, or to get better at a particular technology, then we don't think putting extra tracking in makes sense.
+
+In practice, however, it's pretty much impossible not to track more than you need in this case.
+
+If a project is hosted on GitHub, it automatically provides basic stats (clones / repo visits) provided for all repos.
+
+#### Aggregated tracking
+
+This helps projects validate if the concept of what they're working on is valuable - do people install it? Do they actually use it at all?
+
+It also means they can get a sense of how accessible their software is by looking at how many people work out how to use it. At PostHog, early on, we tracked if people were able to get through the setup process. Many didn't, so we've kept tweaking it - and we still have work to do here :)
+
+With aggregated tracking, a project cannot by definition identify unique users. This means repeat visitors are likely to skew the statistics, and therefore, the understanding of user behavior.
```

Comment:
A technical reader might point out that there are ways to build aggregated tracking while taking into account repeat visitors.

Reply:
I couldn't think of a way of doing this without it falling into the 'anonymized user tracking' bucket below - could you elaborate? Maybe I need to clarify this.

Reply:
Well it really depends on how you classify things, but let's say I anonymize IDs and I'm tracking pageviews.

Your suggestion of aggregate tracking might be something like:

User accesses the page, we literally just bump a counter. 

But you could, for example, without consistently tracking users individually, add that anonymized ID to a set of IDs. 

Then, once the user accesses the page, you have 2 counters: the traditional one that always goes up by 1. A unique visitors counter that checks if the ID exists in the set and bumps the counter otherwise. 

This way, even if I could discover who the anonymized ID is, I can only know that they visited the website at least once, but with no further details. I personally wouldn't call this tracking the user specifically.

And then you can also go into more complex mechanisms like zero-knowledge proofs, where there are systems sparking up that allow people to vote completely anonymously but still ensure each person votes only once.

[Here's a paper on this](https://link.springer.com/chapter/10.1007/978-3-319-70972-7_20).

I do think there's a lot of room for discussion regarding what should go into what bucket though, as this is mostly a discussion of definitions. You could call my first approach tracking for example.

Reply:
> Then, once the user accesses the page, you have 2 counters: the traditional one that always goes up by 1. A unique visitors counter that checks if the ID exists in the set and bumps the counter otherwise.

In this case, I would define this as anonymized user tracking since you have an ID associated with the user.

I\ve clarified the definitions to solve.